### PR TITLE
Refactor and test spec block rules

### PR DESCRIPTION
**Pull request by NikVolf**
This slightly refactors how client processes fork_blocks/bad_blocks internally, and also puts those under tests

### CODE REVIEW

File: client/src/block_rules.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline client/src/block_rules.rs]:
```suggestion
// Copyright 2020 Parity Technologies (UK) Ltd.
```

**Comment by NikVolf** [comment]
@andresilva?

File: client/src/block_rules.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Client fixed specification rules
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
//! Client fixed chain specification rules
```

File: client/src/block_rules.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+use std::collections::{HashMap, HashSet};
+
+use sp_runtime::{
+	traits::{Block as BlockT, NumberFor, BlockHashFor},
+};
+
+use sc_client_api::{ForkBlocks, BadBlocks};
+
+/// Specification rules lookup result.
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
/// Chain specification rules lookup result.
```

File: client/src/block_rules.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+pub enum LookupResult<B: BlockT> {
+	/// Specification rules do not contain any special rules about this block
+	NotSpecial,
+	/// The bock is known to be bad and should not be imported
+	KnownBad,
+	/// There is a fork expected to different fork
```
Comment (andresilva) [inline client/src/block_rules.rs]: There is a specified canonical block hash for the given height.
```suggestion
	/// There is a specified canonical block hash for the given height
```

File: client/src/block_rules.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	ExpectedForkTo(BlockHashFor<B>)
+}
+
+/// Block special rules.
+///
+/// This holds known bad blocks and known good forks, and
+/// usually part of the chain spec.
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
/// is usually part of the chain spec.
```

File: client/src/block_rules.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+pub enum LookupResult<B: BlockT> {
+	/// Specification rules do not contain any special rules about this block
+	NotSpecial,
+	/// The bock is known to be bad and should not be imported
+	KnownBad,
+	/// There is a fork expected to different fork
+	ExpectedForkTo(BlockHashFor<B>)
+}
+
+/// Block special rules.
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
/// Chain-specific block filtering rules.
```

File: client/src/block_rules.rs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+use sc_client_api::{ForkBlocks, BadBlocks};
+
+/// Specification rules lookup result.
+pub enum LookupResult<B: BlockT> {
+	/// Specification rules do not contain any special rules about this block
+	NotSpecial,
+	/// The bock is known to be bad and should not be imported
+	KnownBad,
+	/// There is a fork expected to different fork
+	ExpectedForkTo(BlockHashFor<B>)
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
	Expected(B::Hash)
```

File: client/src/block_rules.rs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+/// Block special rules.
+///
+/// This holds known bad blocks and known good forks, and
+/// usually part of the chain spec.
+pub struct BlockRules<B: BlockT> {
+	bad: HashSet<BlockHashFor<B>>,
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
	bad: HashSet<B::Hash>,
```

File: client/src/block_rules.rs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	forks: HashMap<NumberFor<B>, BlockHashFor<B>>,
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
	forks: HashMap<NumberFor<B>, B::Hash>,
```

File: client/src/block_rules.rs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+		fork_blocks: ForkBlocks<B>,
+		bad_blocks: BadBlocks<B>,
+	) -> Self {
+		Self {
+			bad: bad_blocks.unwrap_or(HashSet::new()),
+			forks: fork_blocks.unwrap_or(vec![]).into_iter().collect(),
+		}
+	}
+
+	pub fn lookup(&self, number: NumberFor<B>, hash: &BlockHashFor<B>) -> LookupResult<B> {
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion
	pub fn lookup(&self, number: NumberFor<B>, hash: &B::Hash) -> LookupResult<B> {
```

File: client/src/client.rs [hunk 11]
```diff
@@ -3002,6 +2993,107 @@ pub(crate) mod tests {
...
+
+			if !record_only {
+				assert_eq!(client.check_block(params).unwrap(), ImportResult::KnownBad);
+			}
+		}
+
+		let mut known_bad = HashSet::new();
+		let mut fork_rules = Vec::new();
+
+		// records what bad_blocks and fork_blocks hashe should be
```
Comment (andresilva) [inline client/src/client.rs]:
```suggestion
		// records what bad_blocks and fork_blocks hashes should be
```

File: primitives/runtime/src/traits.rs [hunk 12]
```diff
@@ -636,6 +636,8 @@ pub type NumberFor<B> = <<B as Block>::Header as Header>::Number;
 pub type DigestFor<B> = Digest<<<B as Block>::Header as Header>::Hash>;
 /// Extract the digest item type for a block.
 pub type DigestItemFor<B> = DigestItem<<<B as Block>::Header as Header>::Hash>;
+/// Extract block hash.
+pub type BlockHashFor<B> = <B as Block>::Hash;
```
Comment (andresilva) [inline primitives/runtime/src/traits.rs]: I don't think this is necessary, in this case we're only doing one projection so the type will always be simple to write (just `B::Hash`).

File: test-utils/runtime/client/src/lib.rs [hunk 13]
```diff
@@ -140,7 +140,9 @@ impl substrate_test_client::GenesisInit for GenesisParameters {
 }
 
 /// A `TestClient` with `test-runtime` builder.
-pub type TestClientBuilder<E, B> = substrate_test_client::TestClientBuilder<E, B, GenesisParameters>;
+pub type TestClientBuilder<E, B> = substrate_test_client::TestClientBuilder<
+	substrate_test_runtime::Block, E, B, GenesisParameters
```
Comment (andresilva) [inline test-utils/runtime/client/src/lib.rs]: I'd format this as one type per line (just an opinion though, not enforcing).

**Comment by NikVolf** [comment]
@andresilva  all addressed, thanks!

File: client/src/block_rules.rs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	pub fn new(
+		fork_blocks: ForkBlocks<B>,
+		bad_blocks: BadBlocks<B>,
+	) -> Self {
+		Self {
+			bad: bad_blocks.unwrap_or(HashSet::new()),
+			forks: fork_blocks.unwrap_or(vec![]).into_iter().collect(),
+		}
+	}
+
```
Comment (andresilva) [inline client/src/block_rules.rs]:
```suggestion

	/// Check if there's any rule affecting the given block.
```

**Pull request merged**